readme.md
# manablox-service-express

express.js Server - Service module for manablox





CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.6.6](https://github.com/manablox/manablox-service-express/compare/v1.6.5...v1.6.6) (2020-01-25)


### Bug Fixes

* forgotten bodyParser config ([5f412f0](https://github.com/manablox/manablox-service-express/commit/5f412f0de7bdd9d8feaaf4bf0094ac888c35d106))

### [1.6.5](https://github.com/manablox/manablox-service-express/compare/v1.6.4...v1.6.5) (2020-01-25)


### Bug Fixes

* config var name fix ([efee1f2](https://github.com/manablox/manablox-service-express/commit/efee1f277a579c73d9afbe9645f79f24d832f8be))

### [1.6.4](https://github.com/manablox/manablox-service-express/compare/v1.6.3...v1.6.4) (2020-01-25)


### Refactorings

* added default config ([6a7c8ee](https://github.com/manablox/manablox-service-express/commit/6a7c8eeac2aa11513b42e2eff34d724031bd85ce))

### [1.6.3](https://github.com/manablox/manablox-service-express/compare/v1.6.2...v1.6.3) (2020-01-19)


### Bug Fixes

* fixed ignored dynamic added routes ([3def512](https://github.com/manablox/manablox-service-express/commit/3def5126e00e7d6a4c3f6599459dfe26b83d1750))

### [1.6.2](https://github.com/manablox/manablox-service-express/compare/v1.6.1...v1.6.2) (2020-01-19)

### [1.6.1](https://github.com/manablox/manablox-service-express/compare/v1.6.0...v1.6.1) (2020-01-19)

## [1.6.0](https://github.com/manablox/manablox-service-express/compare/v1.5.3...v1.6.0) (2020-01-19)


### Features

* added router ([e53856c](https://github.com/manablox/manablox-service-express/commit/e53856c42be781dbdaa29754c2c2fae77e41d9cb))

### [1.5.3](https://github.com/manablox/manablox-service-express/compare/v1.5.2...v1.5.3) (2020-01-19)

### [1.5.2](https://github.com/manablox/manablox-service-express/compare/v1.5.1...v1.5.2) (2020-01-19)

### [1.5.1](https://github.com/manablox/manablox-service-express/compare/v1.5.0...v1.5.1) (2020-01-04)


### Documentations

* added repository to package ([b2cf78e](https://github.com/manablox/manablox-service-express/commit/b2cf78edf561367f14795349bfa96d3d742c8561))

## [1.5.0](https://github.com/manablox/manablox-service-express/compare/v1.4.1...v1.5.0) (2020-01-04)


### Features

* added more router functionality ([8ea600a](https://github.com/manablox/manablox-service-express/commit/8ea600a5b753a841f4fb436caad5e4b9ffe9324b))

### [1.4.1](https://github.com/manablox/manablox-service-express/compare/v1.4.0...v1.4.1) (2020-01-03)


### Refactorings

* added basic test ([d9e49e4](https://github.com/manablox/manablox-service-express/commit/d9e49e46c859f77e267d10f9b16229ce6d82ade3))

## [1.4.0](https://github.com/manablox/manablox-service-express/compare/v1.3.1...v1.4.0) (2020-01-03)


### Features

* added options to disable express plugins ([23d7080](https://github.com/manablox/manablox-service-express/commit/23d70804d988ae296c0ad2ae026c3e85d92f6a65))

### [1.3.1](https://github.com/manablox/manablox-service-express/compare/v1.3.0...v1.3.1) (2020-01-03)


### Bug Fixes

* readded bodyparser ([66c0212](https://github.com/manablox/manablox-service-express/commit/66c0212f8e3a1eb6f4ce6ddda2e34f08ff33e7ae))

## [1.3.0](https://github.com/manablox/manablox-service-express/compare/v1.2.0...v1.3.0) (2020-01-03)


### Features

* added express.use functionality ([f0b7549](https://github.com/manablox/manablox-service-express/commit/f0b754903aa38bf6a9a3e4152e58f28d0bfb0536))

## [1.2.0](https://github.com/manablox/manablox-service-express/compare/v1.1.0...v1.2.0) (2020-01-03)


### Features

* added class export and start function ([887e131](https://github.com/manablox/manablox-service-express/commit/887e13151a321aef0def276a3e984b3c435223f5))

## [1.1.0](https://github.com/manablox/manablox-service-express/compare/v1.0.4...v1.1.0) (2020-01-03)


### Features

* added more options ([e8e4604](https://github.com/manablox/manablox-service-express/commit/e8e4604d95a6178a793ad11cf7a8aa01adcd1bd8))

### [1.0.4](https://github.com/manablox/manablox-service-express/compare/v1.0.3...v1.0.4) (2020-01-02)


### Bug Fixes

* testing ([45b2e5c](https://github.com/manablox/manablox-service-express/commit/45b2e5c121781ad1938d9b60641bc537480635ec))

### [1.0.3](https://github.com/manablox/manablox-service-express/compare/v1.0.2...v1.0.3) (2020-01-02)


### Bug Fixes

* testing ([0566c40](https://github.com/manablox/manablox-service-express/commit/0566c4098404d21dc6964e98bb663a64b6398b5a))

### [1.0.2](https://github.com/manablox/manablox-service-express/compare/v1.0.1...v1.0.2) (2020-01-02)


### Bug Fixes

* testing ([e2dd97b](https://github.com/manablox/manablox-service-express/commit/e2dd97bd415541cbca2635a2f716b888ec25de09))

### 1.0.1 (2020-01-02)


test/index.js
import ExpressService from '../'

const serverConfig = {
    // server ip and port
    ip: process.env.SERVER_IP || '127.0.0.1',
    port: process.env.SERVER_PORT || 3000,


    // autorouting adds automatically routes recursively from the given directory
    autoroute: 'routes', // set to false to disable autorouting


    // security settings (configuration options, see: https://github.com/helmetjs/helmet)
    helmet: {
        disabled: false,
        options: {
            hsts: {
                maxAge: 60 * 60 * 2,
                includeSubDomains: false
            }
        }

    },


    // CORS settings (configuration options, see: https://github.com/expressjs/cors)
    cors: {
        disabled: false,
        options: {
            origin: '*',
            optionsSuccessStatus: 200
        }
    },


    // body parser settings (configuration options, see: https://github.com/expressjs/body-parser)
    bodyParser: {
        disabled: false,
        options: {
            json: {
                limit: '64mb'
            },
            urlencoded: {
                extended: true
            }
        }

    },


    // compression settings (configuration options, see: https://github.com/expressjs/compression)
    compression: {
        disabled: false,
        options: {}
    }
}

const server = new ExpressService(serverConfig)

server.Use('/', async (req, res, next) => {
    res.json('hello')
})

server.Start()


index.js
import cors from 'cors'
import http from 'http'
import helmet from 'helmet'
import express from 'express'
import bodyParser from 'body-parser'
import compression from 'compression'

class ExpressService {
    constructor({
        ip = '127.0.0.1',
        port = 3000,

        autoroute = 'routes',

        helmet = {
            disabled: false,
            options: {
                hsts: {
                    maxAge: 60 * 60 * 2,
                    includeSubDomains: false
                }
            }
        },

        cors = {
            disabled: false,
            options: {
                origin: '*',
                optionsSuccessStatus: 200
            }
        },

        bodyParser = {
            disabled: false,
            options: {
                json: {
                    limit: '64mb'
                },
                urlencoded: {
                    extended: true
                }
            }
        },

        compression = {
            disabled: false,
            options: {}
        }
    }){
        this.config = {
            ip,
            port,
            autoroute,
            helmet,
            cors,
            compression,
            bodyParser
        }

        this.server = null
        this.app = null
        this.logger = null

        this.Initialize()
    }

    SetLogger(logger){

    }

    Initialize(){
        this.app = express()

        if(!this.config.helmet.disabled) this.app.use(helmet(this.config.helmet.options))
        if(!this.config.cors.disabled) this.app.use(cors(this.config.cors.options))
        if(!this.config.compression.disabled) this.app.use(compression(this.config.compression.options))
        if(!this.config.bodyParser.disabled) this.app.use(bodyParser.json(this.config.bodyParser.options.json))
        if(!this.config.bodyParser.disabled) this.app.use(bodyParser.urlencoded(this.config.bodyParser.options.urlencoded))
    }

    Start(){
        this.server = http.createServer(this.app)
        this.server.listen(this.config.port, this.config.ip, () => {
            console.log(`An express server is listening on ${ this.config.ip }:${ this.config.port }`)
        })
    }

    Use(...params){
        return this.app.use(...params)
    }

    Get(...params){
        return this.app.get(...params)
    }

    Post(...params){
        return this.app.post(...params)
    }

    Put(...params){
        return this.app.put(...params)
    }

    Delete(...params){
        return this.app.delete(...params)
    }
}

export default ExpressService
